--- shfs-0.35/shfs/Linux-2.6/inode.c.orig	2006-12-10 12:47:30.000000000 +0100
+++ shfs-0.35/shfs/Linux-2.6/inode.c	2006-12-10 12:57:35.000000000 +0100
@@ -341,12 +341,22 @@
 	return -EINVAL;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
 static struct super_block *
 shfs_get_sb(struct file_system_type *fs_type,
 	    int flags, const char *dev_name, void *data)
 {
 	return get_sb_nodev(fs_type, flags, data, shfs_read_super);
 }
+#else
+static int
+shfs_get_sb(struct file_system_type *fs_type,
+	    int flags, const char *dev_name, void *data,
+	    struct vfsmount *mnt)
+{
+	return get_sb_nodev(fs_type, flags, data, shfs_read_super, mnt);
+}
+#endif
 
 static struct file_system_type sh_fs_type = {
 	.owner		= THIS_MODULE,
